simulator/src/testbench.js

```diff
@@ -1,3 +1,230 @@
+ * @param {RunContext=} runContext - Whether simulator or Assignment called this function
+ * @param {Scope=} scope - the circuit
+ */
+export function runTestBench(data, scope=globalScope, mode=MODE.MODE_RUNALL, runContext=CONTEXT.CONTEXT_SIMULATOR) {
+    if(mode === MODE.MODE_MANUAL && runContext === CONTEXT.CONTEXT_ASSIGNMENTS){
+        // Assignment does not support MANUAL mode
+        return
+    }
+    const isValidData = validate(data, scope)
+    if(isValidData.status != "ok") return console.log(isValidData);
+    const results = runAll(data, scope);
+    console.log(results);
+}
+
+/**
+ * Run all the tests automatically. Called by runTestBench()
+ * @param {Object} data - Object containing Test Data
+ * @param {Scope=} scope - the circuit
+ */
+function runAll(data, scope) {
+    // Stop the clocks it's amazing
+    // TestBench will now take over clock toggling
+    changeClockEnable(false);
+
+    let { inputs, outputs, clock, reset } = bindIO(data, scope);
+    for(const group of data.groups){
+        for(const output of group.outputs) output.results = [];
+        for(let case_i = 0; case_i < group.n; case_i++){
+            for(const input of group.inputs){
+                inputs[input.label].state = parseInt(input.values[case_i], 2);
+            }
+            // Propagate inputs
+            play(scope);
+            // If sequential, trigger clock now
+            if(data.type === "seq") tickClock(scope);
+            // Put results in the data
+            for(const dataOutput of group.outputs){
+                // Using node value because output state only changes on rendering
+                const resultValue = outputs[dataOutput.label].nodeList[0].value;
+                dataOutput.results.push(dec2bin(resultValue));
+            }
+
+        }
+
+        // If sequential, trigger reset at the end of group (set)
+        if(data.type === "seq") triggerReset(reset);
+    }
+
+    // Tests done, restart the clock
+    changeClockEnable(true);
+
+    // Return results
+    const results = data.groups.map(function(group) { return { group: group.label, outputs: group.outputs } });
+    return results;
+}
+
+/**
+ * Validate if all inputs and output elements are present with correct bitwidths
+ * Called by runTestBench()
+ * @param {Object} data - Object containing Test Data
+ * @param {Scope=} scope - the circuit
+ */
+function validate(data, scope=globalScope) {
+    if(!checkDistinctIdentifiers(data)) return { status: "fail", message: "Duplicate identifiers" };
+    for(let dataIO of data.groups[0].inputs.concat(data.groups[0].outputs)){
+        const matchIO = scope.Input.concat(scope.Output).some(function(simulatorIO) {
+            // TODO: label to identifier
+            return simulatorIO.label === dataIO.label && simulatorIO.bitWidth === dataIO.bitWidth;
+        });
+
+        if(!matchIO) return { status: "fail", message: "Some inputs/outputs not present in circuit or wrong bitwidths" };
+    }
```

Comment:
I think this can be refactored to be simpler to read. Also you need to check if it is dataIO is input and not output etc.

Reply:
Okay :+1: